### PR TITLE
Fix AsymmetricAutoencoderKL forward

### PR DESCRIPTION
**Pull request by townwish4git**
# What does this PR do?

Fixes # (issue)
Fix incorrect call to `self.decode()` within `AsymmetricAutoencoderKL.forward()`:
```diff
- dec = self.decode(z, sample, mask).sample
+ dec = self.decode(z, generator, sample, mask).sample
```

related to: [issue#8317](https://github.com/huggingface/diffusers/issues/8317)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sayakpaul @yiyixuxu @DN6

### CODE REVIEW

**Comment by tolgacangoz** [comment]
Thanks for opening this PR!
I wonder what `generator` is used for here?

**Comment by townwish4git** [comment]
> Thanks for opening this PR! I wonder what `generator` is used for here?

The `generator` used here is the one passed into the `AsymmetricAutoencoderKL.forward()`

```diff
def forward(
        self,
        sample: torch.Tensor,
        mask: Optional[torch.Tensor] = None,
        sample_posterior: bool = False,
        return_dict: bool = True,
        generator: Optional[torch.Generator] = None,
    ) -> Union[DecoderOutput, Tuple[torch.Tensor]]:
        ...
-        dec = self.decode(z, sample, mask).sample
+        dec = self.decode(z, generator, sample, mask).sample
        ...
```

**Comment by tolgacangoz** [comment]
But `decode()` function doesn't use it :thinking:.

File: src/diffusers/models/autoencoders/autoencoder_asym_kl.py [hunk 1]
```diff
@@ -176,7 +176,7 @@ def forward(
             z = posterior.sample(generator=generator)
         else:
             z = posterior.mode()
-        dec = self.decode(z, sample, mask).sample
+        dec = self.decode(z, generator, sample, mask).sample
```
Comment (sayakpaul) [inline src/diffusers/models/autoencoders/autoencoder_asym_kl.py:179]: Tend to agree with @tolgacangoz's comments here. It's not used in the `decode()` function. Similar to AutoencoderKL. It's used in the `forward()`:

https://github.com/huggingface/diffusers/blob/413604405fddb4692a8e9a9a9fb6c353d22881ea/src/diffusers/models/autoencoders/autoencoder_kl.py#L445

Comment (sayakpaul) [reply src/diffusers/models/autoencoders/autoencoder_asym_kl.py:179]: I don't think there's a stochastic component in the `decode()` function in the first place. So, I further think there's no need to have `generator` in here too:

https://github.com/huggingface/diffusers/blob/413604405fddb4692a8e9a9a9fb6c353d22881ea/src/diffusers/models/autoencoders/autoencoder_kl.py#L286

@yiyixuxu WDYT here?

**Comment by townwish4git** [comment]
> But `decode()` function doesn't use it 🤔.

I think this is to maintain consistency of `decode()` interfaces across different VAEs. For example, here is a [demo](https://huggingface.co/cross-attention/asymmetric-autoencoder-kl-x-1-5) using `AsymmetricAutoencoderKL`:
```python
...
from diffusers import AsymmetricAutoencoderKL, StableDiffusionInpaintPipeline
...
pipe = StableDiffusionInpaintPipeline.from_pretrained("runwayml/stable-diffusion-inpainting")
pipe.vae = AsymmetricAutoencoderKL.from_pretrained("cross-attention/asymmetric-autoencoder-kl-x-1-5")
pipe.to("cuda")

image = pipe(prompt=prompt, image=image, mask_image=mask_image).images[0]
image.save("image.jpeg")

```

in this case, when origin vae is replaced with `AsymmetricAutoencoderKL`, you don't have to modify the codes that calls `self.vae.decode()` within the pipeline, and you can simply retain the [original code](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py#L1443-L1446) where arguments `generator` is passed in:

```python
image = self.vae.decode(
    latents / self.vae.config.scaling_factor, return_dict=False, generator=generator, **condition_kwargs
)[0]
```

File: src/diffusers/models/autoencoders/autoencoder_asym_kl.py (hunk 1 above)
Comment (townwish4git) [reply src/diffusers/models/autoencoders/autoencoder_asym_kl.py:179]: i think this comment answered the question:
https://github.com/huggingface/diffusers/pull/5694/files#r1387349283

**Comment by tolgacangoz** [comment]
Hmm, you are right. [The `generator` was added to keep interfaces the same as the Consistency Decoder.](https://github.com/huggingface/diffusers/pull/5694#discussion_r1387349283)

**Review by yiyixuxu — APPROVED** [review]
thanks!

File: src/diffusers/models/autoencoders/autoencoder_asym_kl.py (hunk 1 above)
Comment (yiyixuxu) [reply src/diffusers/models/autoencoders/autoencoder_asym_kl.py:179]: @sayakpaul I had the same questions, but their explanation makes sense I think

**Comment by yiyixuxu** [comment]
@tolgacangoz @sayakpaul 
I want to understand a little bit more before merging this PR 
with this demo https://github.com/huggingface/diffusers/pull/8378#issuecomment-2144516387 I think it will currently throw an error so this PR fixes it

is there a different way to use  `AsymmetricAutoencoderKL`? asking because we updated the signature and can potentially break the previous use case that's working. If it is previously not working, we are good

**Comment by townwish4git** [comment]
@yiyixuxu Most use cases call the `AsymmetricAutoencoderKL.encode()` or `decode()` functions, which will be not affected before and after merging.

I'm not sure if there is any `forward()` called directly in the current use cases, and if it exist:
- Before merging:  it wouldn't throw an error, but would have unexpected computed results as [issue#8317](https://github.com/huggingface/diffusers/issues/8317) mentioned
- After merging: get fixed

**Pull request merged**